src/trio/_core/_run.py

```diff
@@ -540,6 +540,11 @@ class CancelScope:
 
     # Constructor arguments:
     _deadline: float = attrs.field(default=inf, kw_only=True, alias="deadline")
+    # use float|None, or math.inf?
+    # or take it as an opportunity to also introduce deadline=None?
+    _relative_deadline: float | None = attrs.field(
+        default=None, kw_only=True, alias="relative_deadline"
+    )
```

Comment:
I don't think `deadline` + `relative_deadline` should be supported: we could deny that using a post-init thing from `attrs`. Maybe someone has a use case for it? I can't imagine anything.

Reply:
cc @Zac-HD because you have opinions.

I think what @Zac-HD proposed of disallowing setting relative deadlines inside the cancel scope would help somewhat, even though I still think that absolute deadlines vs relative deadlines is a weird concept to have before the context manager. (So namely being able to pass both here and being able to set both with their properties)

Reply:
My preferred semantics are:

- `CancelScope` gets a new `._relative_deadline: float` property, with no corresponding init argument
- the `*_after()` scopes create a `CancelScope` with `deadline=inf`, and then set the attribute
- on entry to the `CancelScope`, set the deadline to `min(deadline, _relative_deadline + current_time())`, and then set `_relative_deadline = inf`
  - this means that you get the same semantics for setting `deadline` as any other scope, but "relative deadlines" don't become a semantic thing for users to think about
- it's an error to set the `_relative_deadline` if the scope has ever been entered

The existing deprecation logic is a bit ugly, but seems like the best option available overall.

Reply:
Why is this your preferred semantics? I'm mostly just seeing downsides to having the feature be this hidden, but still able to sneakily introduce confusing behavior if a user is sending around a `CancelScope`. If I have a function that takes a `CancelScope` as a parameter, I'd *have* to introspect the hidden attribute if I wanted to figure out what it would set the deadline to once entered. Maybe that's uncommon logic-wise, but surely people are writing logging statements like `print(f"this cs will time out at {cs.deadline}")` (before entering).

If we don't want to introduce relative deadlines as a primary feature, and avoid any stealthy gotchas, then I strongly think we should introduce a class that transforms into a `CancelScope` - to make sure that typing would catch any attempt at passing/saving the return value of `fail/move_on_after` to something that doesn't expect precisely that, and raise `AttributeError` if attempting to access `deadline`.

Reply:
Oh, the `min` variant also breaks trying to change the deadline, unless we unset `relative_deadline` upon setting `deadline`. So ultimately there's no reason to use `min` and we instead do some conditionals on whether values are `None`/`inf`.

```python
cs = move_on_after(5)
# wait no I want to make the deadline later
cs.deadline = trio.current_time() + 7
with cs:
    print(cs.deadline - trio.current_time()) # prints 5
```
this is regardless of if the time is relative to construction or entering.

and of course
```python
cs = move_on_after(5)
cs.deadline = cs.deadline + 2
with cs:
    print(cs.deadline)  # prints 5, not 7, which it would previously
```

Reply:
I now agree that my proposal upthread is clearly bad 😅